
---
title: "Versatile fastq quality control toolkits and duplication removal for sorted bam files"
author: "Tom Carroll and Wei Wang"
date: "`r format(Sys.Date(), '%m/%d/%Y')`"
package: Rfastp

output:
  BiocStyle::html_document:
     number_sections: yes
     toc: true
vignette: >
  %\VignetteIndexEntry{Rfastp}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\usepackage[utf8]{inputenc}

bibliography: 
  - fastp.bib
  - gencore.bib
---
 
```{r setup, echo=FALSE, results="hide", include = FALSE}
knitr::opts_chunk$set(tidy=FALSE, cache=FALSE,
                      #dev="png",
                      message=FALSE, error=FALSE, warning=TRUE)
options(width=100)
```

# Introduction

The Rfastp package provides an interface to the all-in-one preprocessing for FastQ files toolkit [fastp](https://github.com/OpenGene/fastp)[@10.1093/bioinformatics/bty560] and toolkit removing sequencing duplications and eliminate sequencing errors by generating consensus reads [gencore](https://github.com/OpenGene/gencore)[@Chen2019].

# Installation

Use the `BiocManager` package to download and install the package from
Bioconductor as follows:

```{r getPackage, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("Rfastp")
```

If required, the latest development version of the package can also be installed
from GitHub.

```{r, eval = FALSE}
install.packages('devtools')
devtools::install_github(repo = 'RockefellerUniversity/Rfastp')
```

Once the package is installed, load it into your R session:

```{r}
library(Rfastp)
```


# FastQ Quality Control with rfastp

The package contains three example fastq files, corresponding to a single-end
fastq file, a pair of paired-end fastq files.
```{r}
se_read1 <- system.file("extdata","Fox3_Std_small.fq.gz",package="Rfastp")
pe_read1 <- system.file("extdata","reads1.fastq.gz",package="Rfastp")
pe_read2 <- system.file("extdata","reads2.fastq.gz",package="Rfastp")
output_fastq  <- './rfastp_test'
```

## a normal run for single-end fastq file. 
Rfastp support multiple threads, set threads number by parameter `thread`.
```{r}
se_json_report <- rfastp(read1 = se_read1, outputFastq = './rfastp_test_se',
         thread = 4)
```

## a normal run for paired-end fastq files and specify the adaptor sequence.

```{r}
pe_json_report <- rfastp(read1 = pe_read1, read2 = pe_read2,
         outputFastq = './rfastp_test_pe')
```

## merge paired-end fastq files after QC.

```{r}
pe_json_report <- rfastp(read1 = pe_read1, read2 = pe_read2, merge = TRUE,
         outputFastq = './rfastp_pe_test_unpaired',
         mergeOut = './rfastp_pe_test_merged.fastq.gz')
```

## umi processing

the following example will add prefix string before the UMI sequence in the sequence name. An "_" will be added between the prefix string and UMI sequence. The UMI sequences will be inserted into the sequence name before the first space.

```{r umi}
umi_json_report <- rfastp(read1 = pe_read1, read2 = pe_read2, outputFastq = './rfastp_test_umi',
                          umi = TRUE, umiLoc = "read1", umiLength = 16, umiPrefix = "#", umiNoConnection = TRUE, umiIgnoreSeqNameSpace = TRUE)
```
 
## A quality control example with customized cutoffs. 
Trim poor quality bases at 3' end base by base with quality higher than 5; trim poor quality bases at 5' end by a 29bp window with mean quality higher than 20; disable the polyG trimming, specify the adapter sequence for read1.
```{r}
clipr_json_report <- rfastp(read1 = se_read1, outputFastq = './rfastp_test_clipr',
         disableTrimPolyG = TRUE,
         cutLowQualFront = TRUE,
         cutFrontWindowSize = 29,
         cutFrontMeanQual = 20,
         cutLowQualTail = TRUE,
         cutTailWindowSize = 1,
         cutTailMeanQual = 5,
         minReadLength = 29,
         adapterSequenceRead1 = 'GTGTCAGTCACTTCCAGCGG'
)
```


# Revmove sequencing duplications from sorted bam files with rgencore
The package contains a sorted bam file and a reference file.
```{r}
inputbamfile <-  system.file("extdata", "ex1_sorted.bam", package="Rfastp")
reference <- system.file("extdata", "myreference.fa", package="Rfastp")
outputbamfile <- "ex1_rgencore.bam"
```

run `rgenecore` to remove the duplications from the sorted bam file.
```{r}
rgencore_json_report <- rgencore(inBam=inputbamfile,
         outBam=outputbamfile,
         refFile=reference
)
```

# Miscellaneous helper functions
usage of rfastp:
```{r}
?rfastp
```
usage of rgencore:
```{r}
?rgencore
```
# Acknowledgements
Thank you to Ji-Dung Luo for testing/vignette review/critical feedback, Doug Barrows for critical feedback/vignette review and Ziwei Liang for their support.
# Session info

```{r sessionInfo}
sessionInfo()
```

# References
